sys/usb/usbus/cdc/acm/cdc_acm.c

```diff
@@ -0,0 +1,306 @@
+{
+    for (size_t i = 0; i < len; i++) {
+        tsrb_add_one(&cdcacm->tsrb, buf[i]);
+        if (buf[i] == '\n') {
+            _cdc_acm_flush(cdcacm);
+        }
+    }
+    return len;
+}
+
+/* flush event */
+void _cdc_acm_flush(usbus_cdcacm_device_t *cdcacm)
+{
+    usbus_event_post(cdcacm->usbus, &cdcacm->flush);
+}
+
+int cdc_init(usbus_t *usbus, usbus_cdcacm_device_t *handler,
+             usbus_cdcacm_cb_t cb, uint8_t *buf, size_t len)
+{
+    memset(handler, 0, sizeof(usbus_cdcacm_device_t));
+    handler->usbus = usbus;
+    tsrb_init(&handler->tsrb, buf, len);
+    handler->handler_ctrl.driver = &cdc_driver;
+    handler->cb = cb;
+    handler->state = USBUS_CDCACM_LINE_STATE_DISCONNECTED;
+    usbus_register_event_handler(usbus, (usbus_handler_t*)handler);
+    return 0;
+}
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler)
+{
+    DEBUG("CDC_ACM: intialization\n");
+    usbus_cdcacm_device_t *cdcacm = (usbus_cdcacm_device_t*)handler;
+
+    cdcacm->flush.handler = _handle_flush;
+
+    cdcacm->cdcacm_hdr.next = NULL;
+    cdcacm->cdcacm_hdr.funcs = &_cdcacm_descriptor;
+    cdcacm->cdcacm_hdr.arg = cdcacm;
+
+    /* Configure Interface 0 as control interface */
+    cdcacm->iface_ctrl.class = USB_CLASS_CDC_CONTROL ;
+    cdcacm->iface_ctrl.subclass = USB_CDC_SUBCLASS_ACM;
+    cdcacm->iface_ctrl.protocol = USB_CDC_PROTOCOL_NONE;
+    cdcacm->iface_ctrl.hdr_gen = &cdcacm->cdcacm_hdr;
+    cdcacm->iface_ctrl.handler = handler;
+    /* Configure second interface to handle data endpoint */
+    cdcacm->iface_data.class = USB_CLASS_CDC_DATA ;
+    cdcacm->iface_data.subclass = USB_CDC_SUBCLASS_NONE;
+    cdcacm->iface_data.protocol = USB_CDC_PROTOCOL_NONE;
+    cdcacm->iface_data.hdr_gen = NULL;
+    cdcacm->iface_data.handler = handler;
+
+    /* Create required endpoints */
+    usbus_endpoint_t *ep = usbus_add_endpoint(usbus, &cdcacm->iface_ctrl,
+                                              USB_EP_TYPE_INTERRUPT,
+                                              USB_EP_DIR_IN, 8);
+    ep->interval = 255;
+    usbus_enable_endpoint(ep);
+    ep = usbus_add_endpoint(usbus, &cdcacm->iface_data, USB_EP_TYPE_BULK, USB_EP_DIR_IN, 64);
+    ep->interval = 0;
+    usbus_enable_endpoint(ep);
+    ep = usbus_add_endpoint(usbus, &cdcacm->iface_data, USB_EP_TYPE_BULK, USB_EP_DIR_OUT, 64);
+    ep->interval = 0;
+    usbus_enable_endpoint(ep);
+    usbdev_ep_ready(ep->ep, 0);
+
+    /* Add interfaces to the stack */
+    usbus_add_interface(usbus, &cdcacm->iface_ctrl);
+    usbus_add_interface(usbus, &cdcacm->iface_data);
+
+    usbus_handler_set_flag(handler, USBUS_HANDLER_FLAG_RESET);
+}
+
+static int _setup_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_setuprq_state_t state, usb_setup_t *setup)
+{
+    (void)state;
+    (void)usbus;
+    usbus_cdcacm_device_t *cdcacm = (usbus_cdcacm_device_t*)handler;
+    DEBUG("Request:0x%x\n", setup->request);
```

Comment:
Since this function's name literally contains the words `setup` within something to do with `stdio` I'm just wondering: is this output going anywhere?

Reply:
This question wasn't answered yet, and there are still `DEBUG` statements in this function.

Reply:
This setup (now named control) function handles control messages from the host computer, for example the host signalling that the serial device is open on the host computer or to configure the baud rate.

Depending on whether CDC ACM is used for STDIO, the debug messages either end up in the `trsb` of the CDC ACM STDIO handler and are transmitted to the host at some point or they are immediately written to (uart|rtt|`/dev/null`).

TL;DR: Yes, the output is going somewhere